### DISCUSSION TITLE
upload self written program to Create 3

### DISCUSSION THREAD

**Discussion by arielc-brillianetor** [iRobotEducation/create3_docs#63] [Questions] [answered]
Hi,
I wasn't able to figure out if I can upload a self written program to the create 3 which can activate the internal actions of the robot. I'll give an example. Let's say I write a python code for a client while on my laptop I have a server. I want the client to take the messages I send from the server and activate the robot according to the message. the first message can be "move to coordinates x=10, y=10", the second message can be "turn right 1080 degrees" and the third message can be "return home". 
The communication can be done using the wireless connection. Is the ROS2 already configured to receive messages and act accordingly in real time? Or do I need to upload a program which reads the ROS2 messages and activates the create 3?
Do I need to add any hardware to the robot besides what came in the box?
Thanks for all of the help, I'm new at all of this :)
Ariel

**Comment by shamlian** [comment, marked as answer]
Create 3 implements these [ROS 2 APIs](https://iroboteducation.github.io/create3_docs/api/ros2/) out of the box, and also is compatible with iRobot Coding, including the iRobot Education [Python Web Playground](https://python.irobot.com/). You can just publish requests to the robot (including things like [Navigate to Position](https://iroboteducation.github.io/create3_docs/api/drive-goals/#navigate-to-position)) and it will "just work."

You'd want to add additional hardware, like an SBC, if you think you'd want to use the robot in conjunction with other onboard sensing or actuation. If you just want to use the robot as a turtle, there's no need for any additional hardware -- put the robot on your network, set the RMW, and you're good to go.

It is not possible to upload arbitrary code to the robot.
